streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java

```diff
@@ -108,19 +113,34 @@ private static String getTaskProducerClientId(final String threadClientId, final
             final String threadProducerClientId = getThreadProducerClientId(threadId);
             final Map<String, Object> producerConfigs = config.getProducerConfigs(threadProducerClientId);
 
-            if (processingMode == StreamThread.ProcessingMode.EXACTLY_ONCE_BETA) {
-                producerConfigs.put(ProducerConfig.TRANSACTIONAL_ID_CONFIG, applicationId + "-" + processId);
-                threadProducer = new StreamsProducer(clientSupplier.getProducer(producerConfigs), true, logContext);
+            if (processingMode == EXACTLY_ONCE_BETA) {
+                producerConfigs.put(ProducerConfig.TRANSACTIONAL_ID_CONFIG, transactionalId);
+                threadProducer = new StreamsProducer(clientSupplier.getProducer(producerConfigs), EXACTLY_ONCE_BETA, logContext);
             } else {
-                threadProducer = new StreamsProducer(clientSupplier.getProducer(producerConfigs), false, logContext);
+                threadProducer = new StreamsProducer(clientSupplier.getProducer(producerConfigs), AT_LEAST_ONCE, logContext);
             }
             taskProducers = Collections.emptyMap();
         }
     }
 
+    public void reInitializeThreadProducer() {
+        if (processingMode != EXACTLY_ONCE_BETA) {
+            throw new IllegalStateException("Exactly-once beta is not enabled.");
+        }
+
+        threadProducer.abortTransaction();
```

Comment:
Hmmm... That is a good point. However, now I am wondering if that is the correct thing to do for eos-beta? Insteand the `TaskManager` should control this?

Reply:
We do call `abortTxn` in RecordCollector, so I feel as long as we haven't fixed https://issues.apache.org/jira/browse/KAFKA-9592
the abortTxn call is still not an no-op, and could be dangerous to trigger the ProducerFenced again, just FYI.

Reply:
Agreed, but aborting a TX is done via `StreamProducer#abortTransaction()` and a `ProducerFencedException` would be swallowed within this method -- so it is still seems to be save to try aborting in general?

My comment is about eoa-alpha vs eos-beta: for eos-alpha the record collector can call `abortTransaction()` as it is it's own producer. However, for eos-beta the producer is shared: assume one task is closed but another running task has pending writes: for this case the task that gets closed should not call abortTx or we would loose the pending writes from the other task.

Thoughts?

Reply:
I think there is some broader issue. I would propose to move forward with this PR as-is and address this broader issue (also related to core refactoring) in a follow-up PR.

Hence, I think it's correct to call `abortTransaction()` here.

Reply:
Not sure if all of us are on the same page for this, my understanding is that, in eos-beta, if one task has an error, then we have to abort the whole txn which would affect other tasks. Do you see an alternative solution to avoid this?

Reply:
That sounds correct. However, it seems we don't do this correctly atm (and that is what I want to address in a follow up PR).

Reply:
I think we are still not clear on the safety of calling `abortTxn` here? Are we at the risk of failure to catch the ProducerFenced from here?

Reply:
I think it's safe, because `threadProducer` is a `StreamsProducer` and thus a `ProducerFencedException` would be swallowed.